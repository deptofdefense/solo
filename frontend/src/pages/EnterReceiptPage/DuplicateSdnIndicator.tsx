import React from "react";
import { Alert } from "solo-uswds";

interface DuplicateSdnIndicatorProps {
  isDuplicate: boolean;
}

const DuplicateSdnIndicator: React.FC<DuplicateSdnIndicatorProps> = ({
  isDuplicate
}) => {
  return (
    <div className="margin-bottom-1em overflow-hidden">
      {isDuplicate && (
        <Alert
          status={"error"}
          className="margin-top-1"
          heading={"SDN already exists in table"}
        />
      )}
    </div>
  );
};

export default DuplicateSdnIndicator;
