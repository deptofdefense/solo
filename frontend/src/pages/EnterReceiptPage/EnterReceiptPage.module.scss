.error {
  color: #d54309;
}
