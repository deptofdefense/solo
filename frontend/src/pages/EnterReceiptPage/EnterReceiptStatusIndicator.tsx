import React from "react";
import { Alert } from "@trussworks/react-uswds";
import { LoadingStatus } from "solo-types";

const EnterReceiptStatusIndicator: React.FC<LoadingStatus> = ({
  loading,
  error = false,
  message = ""
}) => (
  <div className="grid-row flex-justify-center flex-12 margin-top-4 tablet:flex-10">
    {!loading && message && (
      <Alert type={error ? "error" : "success"} heading={message} />
    )}
  </div>
);

export default EnterReceiptStatusIndicator;
