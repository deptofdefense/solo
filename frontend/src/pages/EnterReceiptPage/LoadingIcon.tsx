import React from "react";
import classNames from "classnames";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import {
  faSpinner,
  faExclamationCircle,
  faCheck
} from "@fortawesome/free-solid-svg-icons";
import { LoadingStatus } from "solo-types";
import classes from "./EnterReceiptPage.module.scss";

const LoadingIcon: React.FC<LoadingStatus> = ({ loading, error }) => (
  <span>
    <FontAwesomeIcon
      spin={loading}
      icon={loading ? faSpinner : error ? faExclamationCircle : faCheck}
      className={classNames({
        [classes.error]: error
      })}
    />
  </span>
);

export default LoadingIcon;
