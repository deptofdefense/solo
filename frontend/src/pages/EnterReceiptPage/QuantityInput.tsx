import React from "react";
import { Input } from "solo-uswds";

interface QuantityInputProps {
  enteredQuantity?: number;
  onQuantitiyChange: (value: number) => void;
}

const QuantityInput: React.FC<QuantityInputProps> = ({
  enteredQuantity = 0,
  onQuantitiyChange
}) => {
  const onValueChange: React.ChangeEventHandler<HTMLInputElement> = e => {
    const asNum =
      e.currentTarget.value === "" ? 0 : parseInt(e.currentTarget.value);
    if (!Number.isNaN(asNum)) {
      onQuantitiyChange(asNum);
    }
  };

  return (
    <Input
      value={enteredQuantity}
      onChange={onValueChange}
      placeholder="Quantity"
    />
  );
};

export default QuantityInput;
