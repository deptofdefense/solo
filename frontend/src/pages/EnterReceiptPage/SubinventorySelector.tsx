import React from "react";
import { Subinventory, Locator } from "solo-types";
import { Select } from "solo-uswds";

interface SubinventorySelectorProps {
  subinventorys: Omit<Subinventory, "locators">[];
  enteredSubinventoryCode: string;
  onSelectSubinventory: (code: string) => void;
}

export const SubinventorySelector: React.FC<SubinventorySelectorProps> = ({
  subinventorys,
  enteredSubinventoryCode,
  onSelectSubinventory
}) => (
  <Select
    value={enteredSubinventoryCode}
    onChange={e => {
      onSelectSubinventory(e.currentTarget.value);
    }}
    placeholder="Subinventory"
  >
    {subinventorys.map(subinv => (
      <option key={subinv.code} value={subinv.code}>
        {subinv.code}
      </option>
    ))}
  </Select>
);

interface LocatorSelectrProps {
  locators: Locator[];
  enteredLocatorCode: string;
  onSelectLocator: (code: string) => void;
}

export const LocatorSelector: React.FC<LocatorSelectrProps> = ({
  locators,
  enteredLocatorCode,
  onSelectLocator
}) => (
  <Select
    value={enteredLocatorCode}
    onChange={e => {
      onSelectLocator(e.currentTarget.value);
    }}
    placeholder="Locator"
  >
    {locators.map(loc => (
      <option key={loc.code} value={loc.code}>
        {loc.code}
      </option>
    ))}
  </Select>
);
