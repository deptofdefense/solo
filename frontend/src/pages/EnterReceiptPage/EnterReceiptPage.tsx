import React, { useMemo } from "react";
import { Title, Table, SdnInputForm } from "components";
import { Document } from "solo-types";
import { Button } from "solo-uswds";
import createColumns from "./tableColumns";
import EnterReceiptStatusIndicator from "./EnterReceiptStatusIndicator";
import useEnterReceiptDocuments from "./useEnterReceiptDocuments";

const EnterReceiptPage: React.FC = () => {
  const {
    docs,
    addSdn,
    submitAllLoadingStatus,
    submitAll,
    modifyDocument,
    removeDocument
  } = useEnterReceiptDocuments();
  const columns = useMemo(() => createColumns(modifyDocument, removeDocument), [
    modifyDocument,
    removeDocument
  ]);

  return (
    <div className="tablet:margin-x-8 overflow-x-auto">
      <Title>Enter Receipt</Title>
      <EnterReceiptStatusIndicator {...submitAllLoadingStatus} />
      <Table<Document>
        columns={columns}
        data={docs}
        manualPagination={false}
        manualSortBy={false}
      />
      <div className="grid-row flex-align-start flex-justify">
        <SdnInputForm
          onSubmit={addSdn}
          disabled={submitAllLoadingStatus.loading}
        />
        <Button
          onClick={submitAll}
          className="margin-top-1"
          disabled={submitAllLoadingStatus.loading || docs.length < 1}
        >
          Submit All
        </Button>
      </div>
    </div>
  );
};

export default EnterReceiptPage;
