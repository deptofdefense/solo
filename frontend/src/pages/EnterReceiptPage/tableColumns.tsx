import React from "react";
import { Column } from "react-table";
import { Document } from "solo-types";
import LoadingIcon from "./LoadingIcon";

type CreateColumns = () => Column<Document>[];

const createColumns: CreateColumns = () => [
  {
    Header: "Loading",
    Cell: ({
      row: {
        original: { loadingStatus }
      }
    }) => <LoadingIcon {...loadingStatus} />
  },
  {
    Header: "SDN",
    accessor: "sdn"
  },
  {
    Header: "NIIN",
    accessor: "part.nsn",
    id: "niin"
  },
  {
    Header: "Nomenclature",
    id: "nomen",
    accessor: "part.nomen"
  },
  {
    Header: "Quantity",
    id: "quantity",
    accessor: ({ enteredReceivedQty }) => enteredReceivedQty
  },
  {
    Header: "Commodity",
    id: "commodity",
    accessor: "suppadd.desc"
  },
  {
    Header: "Subinventory",
    id: "subinventory",
    accessor: () => "Subinventory"
  },
  {
    Header: "Locator",
    id: "locator",
    accessor: () => "locator"
  }
];

export default createColumns;
