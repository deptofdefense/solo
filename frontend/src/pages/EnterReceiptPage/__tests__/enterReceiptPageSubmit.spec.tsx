import React from "react";
import { render, fireEvent, wait, sleep } from "test-utils";
import { defaultApiResponse } from "solo-types";
import EnterReceiptPage from "../EnterReceiptPage";
import EnterReceiptStatusIndicator from "../EnterReceiptStatusIndicator";

describe("EnterReceiptPage submit all process", () => {
  const fetchMock = jest.fn();
  const defaultDoc = defaultApiResponse.results[0];

  afterEach(() => {
    fetchMock.mockReset();
  });

  const renderAndSearchForDocument = async () => {
    fetchMock.mockResolvedValue(defaultApiResponse);
    const { getByPlaceholderText, getByText, queryByText, ...rest } = render(
      <EnterReceiptPage />,
      {
        authContext: {
          apiCall: fetchMock
        }
      }
    );
    const inpField = getByPlaceholderText("SDN");
    const searchBtn = getByText("Search");
    const submitBtn = getByText("Submit All");
    fireEvent.change(inpField, {
      target: { value: defaultDoc.sdn }
    });
    await wait(() => {
      expect(inpField).toHaveValue(defaultDoc.sdn);
    });
    fireEvent.click(searchBtn);
    await wait(() => {
      expect(fetchMock).toHaveBeenCalled();
      expect(queryByText(defaultDoc.part.nomen)).toBeInTheDocument();
    });
    return {
      getByPlaceholderText,
      getByText,
      queryByText,
      inpField,
      searchBtn,
      submitBtn,
      ...rest
    };
  };

  it("cannot submit all when there are no documents to submit", async () => {
    const { getByText } = render(<EnterReceiptPage />);
    const submitBtn = getByText("Submit All");
    expect(submitBtn).toHaveAttribute("disabled");
  });

  it("posts documents to api and displays success message on submit all", async () => {
    const { submitBtn, queryByText } = await renderAndSearchForDocument();
    fetchMock.mockResolvedValue({});
    fireEvent.click(submitBtn);
    await wait(() => {
      expect(fetchMock).toHaveBeenCalledTimes(2);
      expect(fetchMock.mock.calls[1][0]).toEqual("/documents/d6t");
      expect(fetchMock.mock.calls[1][1]).toMatchObject({
        method: "POST",
        body: JSON.stringify([
          {
            sdn: defaultDoc.sdn,
            status: "D6T",
            quantity:
              defaultDoc.statuses[defaultDoc.statuses.length - 1].projected_qty,
            subinventory: defaultDoc.suppadd.subinventorys[0].code,
            locator: defaultDoc.suppadd.subinventorys[0].locators[0].code
          }
        ])
      });
      expect(
        queryByText("Successfully submitted 1 document(s)")
      ).toBeInTheDocument();
    });
  });

  it("posts documents to api and displays error on submit all failure", async () => {
    const { submitBtn, queryByText } = await renderAndSearchForDocument();
    fetchMock.mockRejectedValue(new Error());
    fireEvent.click(submitBtn);
    await wait(() => {
      expect(queryByText("Something went wrong")).toBeInTheDocument();
    });
  });

  it("cannot submit docs while any of them are still loading", async () => {
    // add a 1 second load time to the api call in order to
    // test that submit cannot be called during that time
    fetchMock.mockImplementation(async () => {
      await sleep(500);
      return defaultApiResponse;
    });
    const { getByPlaceholderText, getByText, queryByText } = render(
      <EnterReceiptPage />,
      {
        authContext: {
          apiCall: fetchMock
        }
      }
    );
    const inpField = getByPlaceholderText("SDN");
    const searchBtn = getByText("Search");
    const submitBtn = getByText("Submit All");
    fireEvent.change(inpField, {
      target: { value: "asdf" }
    });
    fireEvent.click(searchBtn);
    await wait(() => {
      expect(queryByText("asdf")).toBeInTheDocument();
    });
    fireEvent.click(submitBtn);
    await wait(() => {
      expect(
        queryByText(
          "All documents must be successfully loaded before submitting"
        )
      ).toBeInTheDocument();
      expect(queryByText(defaultDoc.sdn)).toBeInTheDocument();
    });
  });

  it("selects first subinventory and first locator by default", async () => {
    const { getByPlaceholderText } = await renderAndSearchForDocument();
    const subinventorySelect = getByPlaceholderText("Subinventory");
    const locatorSelect = getByPlaceholderText("Locator");
    await wait(() => {
      expect(subinventorySelect).toHaveValue(
        defaultDoc.suppadd.subinventorys[0].code
      );
      expect(locatorSelect).toHaveValue(
        defaultDoc.suppadd.subinventorys[0].locators[0].code
      );
    });
  });

  it("selects corresponding locators on selecting different subinventory", async () => {
    const { getByPlaceholderText } = await renderAndSearchForDocument();
    const subinventorySelect = getByPlaceholderText("Subinventory");
    const locatorSelect = getByPlaceholderText("Locator");
    fireEvent.change(subinventorySelect, {
      target: { value: defaultDoc.suppadd.subinventorys[1].code }
    });
    await wait(() => {
      expect(subinventorySelect).toHaveValue(
        defaultDoc.suppadd.subinventorys[1].code
      );
      expect(locatorSelect).toHaveValue(
        defaultDoc.suppadd.subinventorys[1].locators[0].code
      );
    });
  });

  it("can change the quantitiy to a valid number", async () => {
    const { getByPlaceholderText } = await renderAndSearchForDocument();
    const quantityInp = getByPlaceholderText("Quantity");
    expect(quantityInp).toHaveValue(
      defaultDoc.statuses[
        defaultDoc.statuses.length - 1
      ].projected_qty.toString()
    );
    fireEvent.change(quantityInp, {
      target: { value: "30" }
    });
    await wait(() => {
      expect(quantityInp).toHaveValue("30");
    });
    fireEvent.change(quantityInp, {
      target: { value: "not a number" }
    });
    await wait(() => {
      expect(quantityInp).toHaveValue("30");
    });
  });

  it("can select a different locator", async () => {
    const { getByPlaceholderText } = await renderAndSearchForDocument();
    const locatorSelect = getByPlaceholderText("Locator");
    fireEvent.change(locatorSelect, {
      target: { value: defaultDoc.suppadd.subinventorys[0].locators[1].code }
    });
    await wait(() => {
      expect(locatorSelect).toHaveValue(
        defaultDoc.suppadd.subinventorys[0].locators[1].code
      );
    });
  });
});

describe("EnterReceiptPageStatusIndicator component", () => {
  it("renders error alert when status is error", () => {
    const { asFragment } = render(
      <EnterReceiptStatusIndicator loading={false} error={true} />
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it("does not render while loading", async () => {
    const { asFragment } = render(
      <EnterReceiptStatusIndicator loading={true} error={true} />
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it("renders success alert when status is success", async () => {
    const { asFragment } = render(
      <EnterReceiptStatusIndicator loading={true} error={false} />
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
