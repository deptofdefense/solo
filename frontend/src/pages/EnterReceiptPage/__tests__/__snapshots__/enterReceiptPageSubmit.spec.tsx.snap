// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EnterReceiptPageStatusIndicator component does not render while loading 1`] = `
<DocumentFragment>
  <div
    class="grid-row flex-justify-center flex-12 margin-top-4 tablet:flex-10"
  />
</DocumentFragment>
`;

exports[`EnterReceiptPageStatusIndicator component renders error alert when status is error 1`] = `
<DocumentFragment>
  <div
    class="grid-row flex-justify-center flex-12 margin-top-4 tablet:flex-10"
  />
</DocumentFragment>
`;

exports[`EnterReceiptPageStatusIndicator component renders success alert when status is success 1`] = `
<DocumentFragment>
  <div
    class="grid-row flex-justify-center flex-12 margin-top-4 tablet:flex-10"
  />
</DocumentFragment>
`;
